### PR TITLE
Docs: Add mentee selection process

### PR DESCRIPTION
**Pull request by saswatamcode**
This PR adds content for the `Mentee Selection` section i.e, process and recommendations,  based on recent mentor learnings meeting we had with @bwplotka, @fpetkovski & @metonymic-smokey.

Would love to get some feedback on these processes and any additional recommendations that you might want! 🙂 

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: docs/contributing/mentorship.md
Comment (matej-g) [inline docs/contributing/mentorship.md:123]: I'd even say debugging local environment is not strictly part of mentoring, ideally this should be up to the mentee and they should be able to deal with their environment (obviously, mentor should be able to follow their environment during pair programming etc.).

File: docs/contributing/mentorship.md
Comment (matej-g) [inline docs/contributing/mentorship.md:120]: This sounds like condition sine qua non, do we want to have it explicit here?

File: docs/contributing/mentorship.md
Comment (matej-g) [inline docs/contributing/mentorship.md:125]: See above, I would say leave this up to the mentee, but it should not be _too_ exotic. But e.g. if they want to use Emacs we should not be stopping them :joy:.

File: docs/contributing/mentorship.md
Comment (matej-g) [inline docs/contributing/mentorship.md:126]: :+1:

**Review by matej-g — COMMENTED** [review]
Nice to have this in writing :blush:

File: docs/contributing/mentorship.md
Comment (squat) [reply docs/contributing/mentorship.md:123]: I don't think this is strictly necessary and I think it's potentially exclusionary to note it. If mentees want to use windows because it's the machine they have access to, then by all means, let them. They can always use the WSL or something. Alternatively, they can help get our scripts and such working on .ore platforms.

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:116]: This is neither required nor expected. We can note that having prior contributions is favorable because it demonstrates willingness to engage with the community and independence.

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:119]: Let's avoid the potentially judgmental descriptor "proper".
```suggestion
  * The mentee should have access to a stable internet connection over which they can attend regular video calls with mentors.
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:124]: This is confusing. Mentees are not expected to necessarily know Go. There are many frontend student projects as well as potential projects centered around documentation or other community work.

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:125]: This is odd to me. I'm a maintainer and a mentor and I've never used vscode or goland and would not expect any mentee to be required to do so, especially when the tools (like goland) are closed-source and paid.

Comment (saswatamcode) [reply docs/contributing/mentorship.md:125]: So the main motivation for putting this here is to ensure that the mentee uses some editor/IDE that allows them to navigate code and lookup definitions easily across large codebases. We've had some experiences where mentees weren't progressing or getting frequently blocked due to using a very simplistic text editor or something so complex, that they don't know how to operate it well. So they get confused when trying to think of solutions, and have to ask code-level questions frequently. This can be avoided if mentors already know about this while interviewing, and can suggest they switch to something easier, to have a more productive mentorship!

That being said, let me rephrase this line to capture this, and not make it seem like vscode/goland are the only options! 🙂

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md:124]: Will remove!

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md:116]: Let me rephrase here.

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md:125]: Yup, will rephrase as mentioned here: https://github.com/thanos-io/thanos/pull/5920#discussion_r1032731115

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md]: So we didn't interview before, and this condition wasn't being checked, which led to some troublesome experiences. It is sine qua non, but mentioning it explicitly here, so that mentors can be mindful of this as a prerequisite. WDYT? 🙂

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md:123]: Yup, I can rephrase a bit here but generally wanted to mention that having a local environment that the mentee knows how to operate, is beneficial, and helps the mentorship be more productive. I think we can mention linux VM as a fallback if problems due to this arise. We're not excluding any platforms/machines by any means. WDYT? 🙂 

I think @bwplotka can speak more to this as well.

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md]: Have updated this!

File: docs/contributing/mentorship.md
Comment (bwplotka) [reply docs/contributing/mentorship.md:115]: Maybe it's too intrusive, but what if we ask for some proof, forwarded email or signed letter from manager that they are allowed to work on mentorship? (:

File: docs/contributing/mentorship.md
Comment (bwplotka) [inline docs/contributing/mentorship.md:119]:
```suggestion
    * As most of the discussions (meetings/issues/reviews/code comments) are conducted in English, the mentee should be able to communicate well (verbally and in written) in English (first/second language).
```

File: docs/contributing/mentorship.md
Comment (bwplotka) [reply docs/contributing/mentorship.md:124]: Can we put this as the first item? Quite important.

**Review by bwplotka — APPROVED** [review]
Nice, thanks!

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md]: Added line about it. 🙂

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md:112]: Moved to top

File: docs/contributing/mentorship.md
Comment (bwplotka) [inline docs/contributing/mentorship.md:116]:
```suggestion
      * In case the mentee is employed full-time, we recommend asking for some form of proof (letter/forwarded email from manager) that their employer approves of them working on this mentorship is preferable.
```

**Review by bwplotka — DISMISSED** [review]
Thanks! Great, we can always iterate later. @squat are you roughly ok with wording here?

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md:116]: "is preferable" is at the end, so that makes this optional anyway 🙂

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:89]:
```suggestion
We follow these steps when selecting mentees for any given project:
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:91]: Unnecessary comma 
```suggestion
1. After your project has been proposed to a particular mentorship program like LFX or GSoC, ensure that it is discoverable by potential mentees during the application period by sharing it on the relevant GitHub issue. You could also optionally share this on the [#thanos-dev slack channel](https://cloud-native.slack.com/archives/CL25937SP), [Twitter](https://twitter.com/ThanosMetrics), and even during Thanos Contributor Hours.
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:92]: unnecessary commas
```suggestion
2. Wait for the mentee application period to end and log into the relevant portal for the mentorship in order to view mentee applications. You can refer to the relevant documentation that you will receive to familiarize yourself with the portal, application format, and **selection deadlines**. Additional reference: [LFX Mentor docs](https://docs.linuxfoundation.org/lfx/mentorship/mentors/review-mentee-applications-and-tasks), [GSoC mentor guide](https://google.github.io/gsocguides/mentor/).
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:93]: commas and double 'application'
```suggestion
3. Discuss with your co-mentors and shortlist roughly 30% of the mentees that you've found to be suitable by referring to their applications, which can consist of things like CVs, proposals, and "statement of purpose" type questions. Refer to [recommendations](#recommendations) for what to keep in mind while shortlisting.
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:96]:
```suggestion
6. Conduct 20-minute interviews (refer to [recommendations](#recommendations) for what to keep in mind during these interviews) of the shortlisted mentees to get to know them even better and finally select a single person.
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:97]:
```suggestion
7. Send an email to the selected mentee asking for confirmation and if they are happy with this.
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:104]: > hire in the near future.

hire by whom? Thanos isn't a company

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:106]:
```suggestion
We have been able to distill some learnings from several past mentorship programs, which are listed here as recommendations to mentors:
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:115]:
```suggestion
    * Generally, we expect our mentees to be university/college students who would be able to balance both the mentorship and a moderate amount of course load. Commitments larger than this would need to be treated with some caution. Generally, it's only ok to have a full-time job only for more senior engineerswho underst and how to organize their work, be transparent about the mentorship with their employer, etc.
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:118]:
```suggestion
    * However, the contribution should not be a minimal effort one, such as a typo/grammar fix or variable name change. While these are valued contributions, they cannot effectively demonstrate mentee effort. A better indicator would be some change to the code or documentation. Note that this contribution doesn't need to be in a merged or finalized state and can be open.
```

File: docs/contributing/mentorship.md
Comment (squat) [inline docs/contributing/mentorship.md:124]:
```suggestion
    * They should be using a common code editor with IDE functionality for Golang, e.g. VSCode, Goland etc, that they can operate comfortably for the duration of mentorship. The editor should allow them to easily navigate code and lookup definitions across large codebases as this makes it easier to pair program and share knowledge with them.
```

**Review by squat — COMMENTED** [review]
generally looks good. made some more suggestions. phrasing looks pretty good

File: docs/contributing/mentorship.md
Comment (douglascamata) [reply docs/contributing/mentorship.md]: > Unfortunately, we have a visible case where WSL blocked mentees for the majority of mentorship time.

Let's make this better then. I have a WSL2 environment at home that I successfully used for some time to work on Thanos and if there are any related issues I could have a look at them to help.

I 200% agree with @squat's statement.

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md:66]:
```suggestion
This section explains processes around selecting mentees and conducting [retrospectives](#retrospective).
```

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md:104]:
```suggestion
This process is hard and requires time, as it involves holistically looking at mentorship applications and trying to ascertain if they would be able to benefit from this mentorship and ideally become a contributor, maintainer, or even new hire into the respective mentor's company in near future.
```

File: docs/contributing/mentorship.md
Comment (saswatamcode) [reply docs/contributing/mentorship.md]: To reiterate, we aren't excluding mentees from using any machine/OS. Running a linux VM is acceptable in most cases and this is about having really productive mentorship, without the mentor having to spend time, debugging the mentee's local env in case they run into problems and are blocked. 

> Let's make this better then. I have a WSL2 environment at home that I successfully used for some time to work on Thanos and if there are any related issues I could have a look at them to help.

You could update https://thanos.io/tip/contributing/contributing.md/#development with this, but I'd still be against recommending complicated envs to mentees (who are often new to software and open-source), and just defaulting to simple linux VMs.

**Review by squat — DISMISSED** [review]
lgtm

File: docs/contributing/mentorship.md
Comment (fpetkovski) [reply docs/contributing/mentorship.md:112]: Personally I would not put too much emphasis on this and make it a nice to have. Otherwise we are discriminating against people who are a good fit for various other reasons. I would not give this more weight than any other factor, and treat it as more of a tie-breaker.

**Review by bwplotka — APPROVED** [review]

File: docs/contributing/mentorship.md
Comment (matej-g) [reply docs/contributing/mentorship.md]: All good, if there's a rationale for it as you said, then it should be here :+1:

File: docs/contributing/mentorship.md
Comment (douglascamata) [reply docs/contributing/mentorship.md]: It's not complicated, unless we have Service objects using `service.spec.sessionAffinity: ClientIP`. In this case we would need to recompile the WSL 2 kernel, then it becomes complicated.

Otherwise it's literally a matter of installing an app from the Microsoft Store. It integrates easily into Windows itself and any GUI code editor (like, but not limited to, VSCode or Goland).

**Pull request merged**
